### PR TITLE
Use UTF-8 encoding for nginx plugin

### LINKED ISSUE #5337: Nginx plugin crashes when non-ascii configuration file read
## My operating system is (include version):

Arch Linux ARM 4.9.69-1-ARCH

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):

OS package manager (`pacman -S certbot-nginx`)

## I ran this command and it produced this output:

I ran this command:

`sudo certbot --nginx`

and it produced this:

```
Saving debug log to /var/log/letsencrypt/letsencrypt.log
An unexpected error occurred:
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 10453: ordinal not in range(128)
Please see the logfiles in /var/log/letsencrypt for more details.
```

## Certbot's behavior differed from what I expected because:

I found that the error comes from `/etc/nginx/mime.types` which was included in the default nginx package. If there is any non-ascii encoded file in the nginx configuration path, the same error occurs.

## Here is a Certbot log showing the issue:

<details>
<summary>Click to expand</summary>

```
2017-12-18 20:21:34,224:DEBUG:certbot.main:certbot version: 0.20.0
2017-12-18 20:21:34,235:DEBUG:certbot.main:Arguments: ['--nginx']
2017-12-18 20:21:34,243:DEBUG:certbot.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#manual,PluginEntryPoint#nginx,PluginEntryPoint#null,PluginEntryPoint#standalone,PluginEntryPoint#webroot)
2017-12-18 20:21:34,417:DEBUG:certbot.log:Root logging level set at 20
2017-12-18 20:21:34,427:INFO:certbot.log:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2017-12-18 20:21:34,434:DEBUG:certbot.plugins.selection:Requested authenticator nginx and installer nginx
2017-12-18 20:21:34,742:DEBUG:certbot.log:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/bin/certbot", line 11, in <module>
    load_entry_point('certbot==0.20.0', 'console_scripts', 'certbot')()
  File "/usr/lib/python3.6/site-packages/certbot/main.py", line 861, in main
    return config.func(config, plugins)
  File "/usr/lib/python3.6/site-packages/certbot/main.py", line 685, in run
    installer, authenticator = plug_sel.choose_configurator_plugins(config, plugins, "run")
  File "/usr/lib/python3.6/site-packages/certbot/plugins/selection.py", line 189, in choose_configurator_plugins
    authenticator = installer = pick_configurator(config, req_inst, plugins)
  File "/usr/lib/python3.6/site-packages/certbot/plugins/selection.py", line 25, in pick_configurator
    (interfaces.IAuthenticator, interfaces.IInstaller))
  File "/usr/lib/python3.6/site-packages/certbot/plugins/selection.py", line 77, in pick_plugin
    verified.prepare()
  File "/usr/lib/python3.6/site-packages/certbot/plugins/disco.py", line 248, in prepare
    return [plugin_ep.prepare() for plugin_ep in six.itervalues(self._plugins)]
  File "/usr/lib/python3.6/site-packages/certbot/plugins/disco.py", line 248, in <listcomp>
    return [plugin_ep.prepare() for plugin_ep in six.itervalues(self._plugins)]
  File "/usr/lib/python3.6/site-packages/certbot/plugins/disco.py", line 130, in prepare
    self._initialized.prepare()
  File "/usr/lib/python3.6/site-packages/certbot_nginx/configurator.py", line 156, in prepare
    self.parser = parser.NginxParser(self.conf('server-root'))
  File "/usr/lib/python3.6/site-packages/certbot_nginx/parser.py", line 35, in __init__
    self.load()
  File "/usr/lib/python3.6/site-packages/certbot_nginx/parser.py", line 42, in load
    self._parse_recursively(self.config_root)
  File "/usr/lib/python3.6/site-packages/certbot_nginx/parser.py", line 63, in _parse_recursively
    self._parse_recursively(subentry[1])
  File "/usr/lib/python3.6/site-packages/certbot_nginx/parser.py", line 53, in _parse_recursively
    trees = self._parse_files(filepath)
  File "/usr/lib/python3.6/site-packages/certbot_nginx/parser.py", line 203, in _parse_files
    parsed = nginxparser.load(_file)
  File "/usr/lib/python3.6/site-packages/certbot_nginx/nginxparser.py", line 122, in load
    return loads(_file.read())
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 10453: ordinal not in range(128)
2017-12-18 20:21:34,759:ERROR:certbot.log:An unexpected error occurred:
```

</details>

### PR DESCRIPTION
**Pull request by schoen**
Based on and supersedes #5341.

Fixes #5337.

### CODE REVIEW

**Comment by schoen** [comment]
@ohemorange, I've just merged this with the current master branch and I think @g6123 had actually satisfied all of our outstanding substantive requests on the PR. Can you think of other things that need to be changed or improved?

File: certbot-nginx/certbot_nginx/parser.py [hunk 1]
```diff
@@ -412,10 +417,13 @@ def duplicate_vhost(self, vhost_template, remove_singleton_listen_params=False,
 def _parse_ssl_options(ssl_options):
     if ssl_options is not None:
         try:
-            with open(ssl_options) as _file:
+            with io.open(ssl_options, "r", encoding="utf-8") as _file:
                 return nginxparser.load(_file)
         except IOError:
             logger.warning("Missing NGINX TLS options file: %s", ssl_options)
+        except UnicodeDecodeError:
```
Comment (ohemorange) [inline certbot-nginx/certbot_nginx/parser.py]: This code needs coverage

**Comment by ohemorange** [comment]
@schoen this PR is so close! when do you think you'll get a chance to finish it up?

**Comment by ohemorange** [comment]
Hey @schoen, any chance we could finish this up?

**Comment by ohemorange** [comment]
Bumping on this

**Comment by g6123** [comment]
@ohemorange @schoen Hi, it's been a long time no see. Happily, now I'm ready to finish this up. Can I make another PR?

**Pull request closed without merging**